src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/ModelBinderFactory.cs

```diff
@@ -0,0 +1,229 @@
+            }
+
+            return result;
+        }
+
+        private class DefaultModelBinderProviderContext : ModelBinderProviderContext
+        {
+            private readonly ModelBinderFactory _factory;
+
+            public DefaultModelBinderProviderContext(
+                ModelBinderFactory factory,
+                ModelBinderFactoryContext factoryContext)
+            {
+                _factory = factory;
+                Metadata = factoryContext.Metadata;
+                BindingInfo = factoryContext.BindingInfo;
+
+                MetadataProvider = _factory._metadataProvider;
+                Stack = new List<KeyValuePair<Key, DelegatingBinder>>();
+            }
+
+            public DefaultModelBinderProviderContext(
+                DefaultModelBinderProviderContext parent,
+                ModelMetadata metadata)
+            {
+                Metadata = metadata;
+
+                _factory = parent._factory;
+                MetadataProvider = parent.MetadataProvider;
+                Stack = parent.Stack;
+
+                BindingInfo = new BindingInfo()
+                {
+                    BinderModelName = metadata.BinderModelName,
+                    BinderType = metadata.BinderType,
+                    BindingSource = metadata.BindingSource,
+                    PropertyBindingPredicateProvider = metadata.PropertyBindingPredicateProvider,
+                };
+            }
+
+            public override BindingInfo BindingInfo { get; }
+
+            public override ModelMetadata Metadata { get; }
+
+            public override IModelMetadataProvider MetadataProvider { get; }
+
+            // Not using a 'real' Stack<> because we want random access to modify the entries.
+            public List<KeyValuePair<Key, DelegatingBinder>> Stack { get; }
+
+            public override IModelBinder CreateBinder(ModelMetadata metadata)
+            {
+                var nestedContext = new DefaultModelBinderProviderContext(this, metadata);
+                return _factory.CreateBinderCore(nestedContext, token: null);
+            }
+        }
+
+        private class NoOpBinder : IModelBinder
+        {
+            public static readonly IModelBinder Instance = new NoOpBinder();
+
+            public Task BindModelAsync(ModelBindingContext bindingContext)
+            {
+                bindingContext.Result = ModelBindingResult.Failed(bindingContext.ModelName);
+                return TaskCache.CompletedTask;
+            }
+        }
+
+        private struct Key : IEquatable<Key>
+        {
+            private readonly ModelMetadata _metadata;
+            private readonly object _token;
+
+            public Key(ModelMetadata metadata, object token)
+            {
+                _metadata = metadata;
+                _token = token;
+            }
+
+            public bool Equals(Key other)
+            {
+                return _metadata.Equals(other._metadata) && object.Equals(_token, other._token);
```

Comment:
Perhaps `ReferenceEquals()` for the tokens? Put another way, does `ParameterInfo` implement a useful `Equals()`?